### PR TITLE
Improve Location Comparison

### PR DESCRIPTION
**Pull request by sovdeeth**
### Description

Recently, @Fusezion and Orestaz on the SkUnity discord discovered that it was quite difficult to compare the location of a wall-mounted button to a set of created locations due to how Skript displays yaw.

Since Minecraft and Bukkit store yaws from -180 to 180, a yaw of -90 in a Location is actually output as 270 through ExprYawPitch.
This causes a new location created with that yaw to not match the original location.

This is because Bukkit compares the exact bits of the location fields in Location#equals(). It also means that a location with a pitch of 0 is not equal to the exact same location with a pitch of -0.0. This is great for checking if they're exactly equal, but for our purposes I believe we should be less strict.

This PR adds a Location - Location comparator that uses the toVector method and the Vector#equals() method to compare xyz without floating point issues, checks the worlds for equality, and normalizes the yaw and pitch before checking those for equality.

This should be consistent with how the locations are actually expressed ingame and should cut down on issues with location comparison.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/util/LocationUtils.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	 */
+	public static Location normalize(Location location) {
+		location.setPitch(Location.normalizePitch(location.getPitch()));
+		location.setYaw(Location.normalizeYaw(location.getYaw()));
+
+		if (location.getYaw() == -0.0) location.setYaw(0);
+		if (location.getPitch() == -0.0) location.setPitch(0);
+		if (location.getX() == -0.0) location.setX(0);
+		if (location.getY() == -0.0) location.setY(0);
+		if (location.getZ() == -0.0) location.setZ(0);
```
Comment (Fusezion) [inline src/main/java/ch/njol/skript/util/LocationUtils.java:45]: Unsure if it'll matter too much but I know we complain about in syntax classses.
```suggestion
		if (location.getYaw() == -0.0)
			location.setYaw(0);
		if (location.getPitch() == -0.0)
			location.setPitch(0);
		if (location.getX() == -0.0)
			location.setX(0);
		if (location.getY() == -0.0)
			location.setY(0);
		if (location.getZ() == -0.0)
			location.setZ(0);
```

**Review by Fusezion — CHANGES_REQUESTED** [review]
One minor suggestion, besides this looks alright.

**Comment by Fusezion** [comment]
> I've added two methods to compare locations without yaw/pitch, one with a tolerance value to allow rough comparisons. I think these may be useful for future syntaxes like `if loc1 is roughly equal to loc2` or something similar, but they're not currently used and it may be best to just add them when we actually need them

I believe this is fine to stay, I said it on discord but I'll mention it here too. This would allow us to easily add a block to location comparison where `0.5` can be excluded, not everyone remembers to add it and doesn't make sense to be strict on this comparison.

**Review by Fusezion — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/data/DefaultComparators.java [hunk 2]
```diff
@@ -630,6 +632,25 @@ public boolean supportsOrdering() {
 				return false;
 			}
 		});
+
+		// Location - Location
+		Comparators.registerComparator(Location.class, Location.class, new Comparator<Location, Location>() {
+			@Override
+			public Relation compare(Location location1, Location location2) {
+				// check normal equality first
+				if (location1.equals(location2)) return Relation.EQUAL;
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/classes/data/DefaultComparators.java:641]:
```suggestion
				if (location1.equals(location2))
					return Relation.EQUAL;
```

**Review by AyhamAl-Ali — APPROVED** [review]
Noice ⚡

**Pull request merged**